Feat/signin

ahmedyhia123 opened the pull request:
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

✅ Added Sign In feature:

- Designed the Sign In screen UI using Flutter widgets.
- Integrated Sign In API using Retrofit & Dio.
- Used Cubit for managing authentication state.
- Stored auth token securely using SharedPreferences after successful login.

## Related Tickets & Documents


## QA Instructions, Screenshots, Recordings

To test this feature:

1. Go to the Sign In screen.
2. Enter email and password:
   - Try with valid credentials (expect success and navigation).
   - Try with invalid credentials (expect error message).
3. After login:
   - Ensure token is stored correctly in SharedPreferences.
   - Test if token can be retrieved for future requests.

(Optional Screenshot)
![Sign In UI]
<img src="https://github.com/user-attachments/assets/4613090f-5ca5-4235-9833-443ea038c244" width="300" alt="Login UI Screenshot" />




### UI accessibility checklist

- [x] Used semantic Flutter widgets (TextField, ElevatedButton, etc)
- [x] Form fields validated and keyboard accessible
- [x] Checked contrast and error states manually

## Added/updated tests?

- [ ] Yes
- [x] No,

## [optional] Are there any post deployment tasks we need to perform?

- None (Token already saved via SharedPreferences)

## [optional] What gif best describes this PR or how it makes you feel?

![All set!](https://media.giphy.com/media/111ebonMs90YLu/giphy.gif)

ahmedyhia123 committed:
feat: add sign-in API and response models

ahmedyhia123 committed:
feat: Implement login functionality with SignIn use case and Cubit state management

- Added SignIn use case to handle user authentication.
- Created SigninCubit for managing login state and actions.
- Developed LoginScreen with form validation and user feedback.
- Integrated API service for sign-in requests.
- Updated dependency injection configuration for new classes.
- Enhanced localization support for login-related strings.
- Added error handling for server responses during sign-in.
- Created user entity to manage user data after authentication.
- Implemented login form with loading state and error messages.

ahmedyhia123 committed:
feat: Add email validation messages and improve error handling in login process

ahmedyhia123 committed:
feat: Implement TokenService for managing authentication tokens and update AuthRemoteDataSource to utilize it

ahmedyhia123 committed:
feat: Add navigation to Forget Password page from Login form

ahmedyhia123 committed:
feat: Add pull request template to guide contributors on submission process and requirements

ali72-20 commented on lib/core/services/token_service.dart:
```diff
@@ -0,0 +1,19 @@
+
+@injectable
+class TokenService {
+    final SharedPreferences _prefs;
```
token is very sensitive data for user and shared pref create a directry on android files and any one can find this token and get all user data

so we need to hash or secure this token then you need to search for `flutter secure storage package`

ali72-20 commented on lib/features/auth/api/client/api_servises.dart:
```diff
@@ -14,6 +16,7 @@ abstract class ApiServises {
   factory ApiServises(Dio dio) = _ApiServises;
-
+  @POST(Endpoints.signinEndPoint)
+  Future<SigninRespones> signIn(@Body() Map<String, dynamic> body);
```
using request model more better than using Map for body 
use request body with serlization to make a request body

ahmedyhia123 replied:
i can't get it , please can you explain more?

ali72-20 commented on lib/features/auth/data/source/auth_api_service.dart:
```diff
@@ -1,4 +1,10 @@
+import 'package:exam_app/features/auth/domin/entities/user_entity.dart';
+
 abstract class AuthRemoteDataSource {
+  Future<Either<Failure, UserEntity>> signIn(SigninReqParams params);
```
Either is good solution but we can handle api calls without using any additional packages

ali72-20 commented on lib/features/auth/domin/repository/auth_repo.dart:
```diff
@@ -1,7 +1,12 @@
 abstract class AuthRepo {
 
-  
+  Future<Either<Failure, UserEntity>> signIn(SigninReqParams params);
```
we can handle api calls without using any additional packages you need to search for safe api call 

[[](url)](https://github.com/ali72-20/safe_api_call_demo.git)

you can check the code demo

ali72-20 commented on lib/features/auth/domin/usecases/signin.dart:
```diff
@@ -0,0 +1,17 @@
+  AuthRepoImpl _authRepoImpl;
+  SigninUsecase(this._authRepoImpl);
+  @override
+  Future<Either<Failure, UserEntity>> call({SigninReqParams? params}) async {
```
we can handle api calls without using any additional packages you need to search for safe api call

[](https://github.com/ali72-20/safe_api_call_demo.git)

you can check the code demo

ali72-20 commented on lib/features/auth/presentation/views/screen/login_page.dart:
```diff
@@ -1,15 +1,39 @@
   Widget build(BuildContext context) {
-    return const Placeholder();
+    return BlocProvider<SigninCubit>(
+      create: (context) => getIt.get<SigninCubit>(),
```
avoid using `getIt.get<SigninCubit>()` inside the build method this will lead for every build will create a new instence of view model 

// solution 

```
class LoginScreen extends StatelessWidget {
  LoginScreen({super.key});


final viewModel = getIt.get<SigninCubit>();
@override
  Widget build(BuildContext context) {
    return BlocProvider<SigninCubit>(
      create: (context) => viewModel,
}
```

this will be safe 

or if you using statefull widget you can make the cubit lat 
and use getIt inside the init state that widget life cycle method whiched called only once on start of creating the widget

ali72-20 commented on lib/features/auth/presentation/views/screen/login_page.dart:
```diff
@@ -1,15 +1,39 @@
+      child: Scaffold(
+        appBar: AppBar(
+          leading: IconButton(
+            onPressed: () {},
```
if we had onPress or onTap and it's related for task for next spring or coming task make a comment not notify the reviewer

ali72-20 reviewed (commented):
good job it's just a beginning next time will be more perfect 
keep going

ahmedyhia123 committed:
feat: add database module for shared preferences and secure storage

- Implemented DatabaseModule to provide SharedPreferences and FlutterSecureStorage.
- Added preResolve annotation for asynchronous initialization of SharedPreferences.
- Configured secureStorage as a lazy singleton for efficient resource management.

ahmedyhia123 committed:
feat: Refactor authentication flow to use ApiResult and implement SignInRequest model
